pkg/kine/drivers/generic/generic.go

```diff
@@ -381,6 +395,142 @@ func (d *Generic) queryRowPrepared(ctx context.Context, txName, sql string, prep
 	return r
 }
 
+func (d *Generic) InitializeWriteBatching(ctx context.Context) {
+
+	if !d.BatchingEnabled {
+		return
+	}
+
+	if d.BatchingInterval < 10*time.Millisecond {
+		logrus.WithFields(logrus.Fields{"interval": d.BatchingInterval, "queue_size": d.BatchingMaxQueries}).Print("Interval is less than 10ms, disabling batching of insert queries")
+		return
+	}
+
+	if d.BatchingMaxQueries < 5 {
+		logrus.WithFields(logrus.Fields{"interval": d.BatchingInterval, "queue_size": d.BatchingMaxQueries}).Print("Queue size is less than 5, disabling batching of insert queries")
+		return
+	}
+
+	logrus.Warn("Batching of write queries is currently experimental and MUST NOT be used in production. Expect data loss.")
+	logrus.WithFields(logrus.Fields{"interval": d.BatchingInterval, "queue_size": d.BatchingMaxQueries}).Print("Enabling batching of insert queries")
+	d.flushCh = make(chan struct{})
+	d.addToBatchQueueCh = make(chan BatchedInsert)
+	d.removeFromBatchQueueCh = make(chan int)
+	d.batchingQueue = make([]BatchedInsert, 0, d.BatchingMaxQueries)
+	d.batchingCond = sync.Cond{L: &d.batchingMu}
+
+	go d.watchBatchQueue(ctx)
+	go d.batchQueueWatcher(ctx)
+
+}
+
+func (d *Generic) batchQueueWatcher(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case item := <-d.addToBatchQueueCh:
+			d.batchingCond.L.Lock()
+			if len(d.batchingQueue) >= d.BatchingMaxQueries {
+				d.flushCh <- struct{}{}
+				d.batchingCond.Wait()
+			}
+			d.batchingQueue = append(d.batchingQueue, item)
+			d.batchingCond.L.Unlock()
+		}
+	}
+}
+
+func (d *Generic) processBatchedInserts(ctx context.Context, queue []BatchedInsert) ([]int64, error) {
+	valueStrings := make([]string, 0, len(queue))
+	// len(queue) * argument_count per insert
+	valueArgs := make([]interface{}, 0, len(queue)*8)
+
+	for _, insertItem := range queue {
+		valueStrings = append(valueStrings, "(?, ?, ?, ?, ?, ?, ?, ?)")
+		valueArgs = append(valueArgs, insertItem.args...)
+	}
+
+	// Maybe move this into the Generic struct like other queries
+	// Although since we are dynamically generating this we'd only store the template string there
+	stmt := fmt.Sprintf("INSERT INTO kine(name, created, deleted, create_revision, prev_revision, lease, value, old_value) VALUES%s", strings.Join(valueStrings, ","))
+
+	prepared, err := d.DB.Prepare(stmt)
+	if err != nil {
+		logrus.WithError(err).Error("Process batch queue error")
+		return nil, err
+	}
+
+	row, err := d.executePrepared(ctx, "batch_insert_last_insert_id_sql", stmt, prepared, valueArgs...)
```

Comment:
If we get two insert requests with the same key we fail with:
```
Οκτ 21 20:27:17 aurora microk8s.daemon-k8s-dqlite[234689]: time="2023-10-21T20:27:17+03:00" level=error msg="Process batch queue error" error="exec (try: 0): UNIQUE constraint failed: kine.prev_revision, kine.name"
```

Reply:
The prev_revision should also be unique, probably

Reply:
This seems tricky, initial plan would be maybe try to adjust the `prev_revision` and increment by `i` in `range queue`. But multiple queue processing cycles seems to come with the same revision. The reason being the revision is fetched from the table before they are processed. Maybe keeping the revision in memory instead of fetching from the db and incrementing accordingly might help although I'm not sure how it would play out on the HA scenario.